### ISSUE TITLE
Feature: Add the capability to pin cpu/s to docker service create in the same way that this can be done with docker run --cpu-set=

### ISSUE THREAD

**Issue by piersharding** [docker/docker#30477]
Hi -

I am working on an HPC project that would need to achieve CPU pinning for containers so that specific tasks run on the appropriate CPU as well as speak directly to the host network adapter (a topic covered well in  #25303). This could be achieved by having a head node task (perhaps on the swarm manager node) that uses docker swarm to run an agent container on each compute node launched as a service and then each agent container could then launch the appropriately configured task containers locally using docker run --cpuset-cpus="1,2" etc. 
However, it would be much cleaner/nicer to manage the entire deployment process through docker swarm so that orchestration features would not need to be implemented in the agent nodes - something that would simplify the entire orchestration control structure.

On this basis,  please consider the inclusion of the --cpuset-cpus parameter.

Thanks,
Piers Harding.

**Comment by thaJeztah** [comment]
Could you explain why these tasks have to be pinned to specific CPU's, and `--limit-cpu` / `--reserve-cpu` is not sufficient for your use case?


also, /cc @crosbymichael

**Comment by piersharding** [comment]
Hi -

Yes - I think for most use cases they would be sufficient, and I had not found enough background on them to understand them (I'm hoping that http://stackoverflow.com/questions/38200028/docker-service-limits-and-reservations is correct). This should do what I need - thanks.
However, I'm wondering if there are other cases in heterogeneous systems where CPUs with different instruction sets, or co-processors are addressable by CPU id only (instead of --device as for GPGPUs etc)?

Cheers,
Piers Harding.

**Comment by barorion** [comment]
@thaJeztah, processor affinity is sometimes very important for [achieving high (any predictable) performance](http://highscalability.com/blog/2012/3/29/strategy-exploit-processor-affinity-for-high-and-predictable.html). It is critical for projects like [Aeron](https://github.com/real-logic/Aeron) and [MAX Disruptor](https://github.com/LMAX-Exchange/disruptor). 
Having support for processor affinity at swarm mode level is highly desirable in some use cases, e.g. when running Aeron's media driver as a service.

**Comment by yonger1516** [comment]
@thaJeztah there is still another problem to be resolved: assume we have add processor affinity feature into Docker, we still can't use it well, because we can't allocate one specific replica to a specific NUMA node. 
Certainly, you may say we can bind all replicas of this service to one NUMA node, and then bind another service to anther NUMA node, but it will introduce other problems, like each service has different level of HW resource requirements so that we can't utilize a physical host efficiently in some cases.
So I persist to be think Docker Swarn will allow assign different option value for one of replica:
docker service update --cpuset=0  <replication id>

**Comment by barorion** [comment]
any plans for this?

**Comment by qlyoung** [comment]
+1. Many workloads that would like to run in a container also would like to maintain their CPU affinity.

**Comment by erfansahaf** [comment]
Any news?
It's 2020 guys, just tell something if you have any plan for this.
Thanks.

**Comment by thaJeztah** [comment]
AFAIK, I haven't seen someone working on this, but feel free to work on a design/proposal. Things to be taking into account would be that not all nodes in the cluster may be equal (so nodes could have a different number of CPUs/cores); likely the scheduler would have to take that into account somehow (so that a task for a service having `--cpuset-cpus="3,4"` would only land on a (worker)node that has 3 CPUs.

**Comment by qlyoung** [comment]
@thaJeztah thanks for the reply. Slightly off topic, but as I understand it, it should be possible to set cpu affinity a la `taskset` within the container itself. However, this requires the container to be privileged. And swarm does not support privileged containers either, so this strategy is also not viable with swarm. Have I got this correct?

**Comment by nevgeniev** [comment]
+1 here, just surprised such obvious feature is not implemented yet. out of curiosity, is swarm still being developed?

**Comment by d3matt** [comment]
Just wanted to add my use case: I need to limit both the cpus and mems a process can utilize to eliminate any accidental cross-NUMA memory bandwidth using the --cpuset-cpus AND --cpuset-mems arguments to docker run.

**Comment by jnunezgts** [comment]
Hello,

Docker Swarm should include support for [NUMA ](https://queue.acm.org/detail.cfm?id=2513149)(cpusets-cpus, cpuset-mem). There are many time sensitive applications out there (market data feed handlers, pricing engines) that require this feature. Kubernetes (k8s) [supports this](https://kubernetes.io/docs/tasks/administer-cluster/cpu-management-policies/) with '[CPU Manager](https://kubernetes.io/blog/2018/07/24/feature-highlight-cpu-manager)' since 2018.

Many applications in HFT and Algorithmic trading use NUMA and Docker already leverages this when you use 'docker run', it should be available at the Swarm level.

**Comment by jnunezgts** [comment]
> AFAIK, I haven't seen someone working on this, but feel free to work on a design/proposal. Things to be taking into account would be that not all nodes in the cluster may be equal (so nodes could have a different number of CPUs/cores); likely the scheduler would have to take that into account somehow (so that a task for a service having `--cpuset-cpus="3,4"` would only land on a (worker)node that has 3 CPUs.

[Kubernetes ](https://kubernetes.io/blog/2018/07/24/feature-highlight-cpu-manager/)has this feature.

**Comment by qlyoung** [comment]
> Kubernetes has this feature.

Yes, it does, and this is what made me have to ditch Swarm for k8s. Unfortunately.

**Comment by aleek** [comment]
OK guys. The time has come. Time to put the end on this madness. I am going to develop this feature.

Before we actually start to discuss about the possible drawbacks and corner cases of this feature, I want to develop PoC for the very specific case - when all nodes are equal in terms of resources. This is very easy IMO, just pass the `cpuset-cpus` down to container creation and see what will blow up ;)

Right now I am trying to track down the path from `docker service create --limit-cpu [...] --reserve-cpu [..]` down to container creation.

Limits and reservations are defined in SwarmSpec (api/types/swarm/task.go:96) and they travel down up to (daemon/cluster/services.go:182), the `CreateService` function. This is where I'm currently stuck. **Any idea where it comes up?** I am trying the track down the place where `ContainerSpec` is derived from `SwarmSpec`, because I suppose this will be the place where most of the work will be done.

Best regards
Alex, StreamVX

**Comment by aleek** [comment]
Also, I fell like 80% of you needs the simpliest case - 1 swarm node, or N exact same nodes. Am I correct?

@jnunezgts @d3matt @nevgeniev @qlyoung @piersharding

**Comment by jnunezgts** [comment]
Hello Aleksander,

In our scenario we were not contemplating to have multiple swarm nodes.

--Jose

**Comment by qlyoung** [comment]
In my case the nodes can be heterogeneous. My use case was to inspect the available CPUs and select one to pin against from within the container.

**Comment by carloalbertobarbano** [comment]
+1 for the feature, I would really need it. I am running an HPC clusters with multiple identical nodes, and I would like my containers to be limited to a specific set of cores depending on which GPUs they utilize.

**Comment by aleek** [comment]
> +1 for the feature, I would really need it. I am running an HPC clusters with multiple identical nodes, and I would like my containers to be limited to a specific set of cores depending on which GPUs they utilize.

I have already working PoC. Give me a bit more time, I need to figure out other things, and I'll publish the code.

Alex

**Comment by carloalbertobarbano** [comment]
While waiting for @aleek's patch, we developed a small utility which can run as swarm services and takes care of automatically assigning cpu cores based on the required GPUs (if gpu computing is your use-case). Documentation is lacking, but the idea is very simple, and could perhaps be adapted to different use cases.. https://github.com/EIDOSlab/swarm-cpupin

**Comment by KarstenB** [comment]
One use-case for me is that I have a 96 core build machine, and I want to run 8 agents on them. If I use reservations and limits, `nproc` still returns 96 cores, which some build systems will use to run nproc amounts of jobs. Running 96 jobs, when only having a cpu limit of 8 is of course suboptimal. However with cpusets nproc reports the proper amount of cores available. So in the end, all I want is that nproc reports `max(1, floor(limit-cpu))` amount of cores. I don't necessarily need or want cpu pinning.

But a nice way of doing cpu pinning would be to have cpuset be computable, like this:
`{{-  .Task.ID*8 }}-{{- ((.Task.ID+1)*8)-1 }},64-96`

**Comment by lordraiden** [comment]
Any news about this?

**Comment by andyfu6** [comment]
When docker service can support cpu binding?

This feature is very necessary to make high performance program stable.
